### ISSUE TITLE
Post quantum encryption

### ISSUE THREAD

**Issue by hammad93** [hammad93/time_crypt#15]
The current NIST guidelines rely on the bit strength being greater than the q-bits present in the competing system. If we set it to an arbitrarily high number it should be sufficient for a few months. We can implement their quantum encryption algorithm which allows an increase in bit strength retroactively. However, this means that the initial bit strength and subsequent increases will cause the messages generated with the lower bit strength to still be vulnerable.

**Comment by hammad93** [comment]
We can test to see whether we can generate a quantum key with a TPM module.

**Comment by hammad93** [comment]
https://www.nist.gov/news-events/news/2024/08/nist-releases-first-3-finalized-post-quantum-encryption-standards

**Comment by hammad93** [comment]
- FIPS 203, [Module-Lattice-Based Key-Encapsulation Mechanism Standard](https://csrc.nist.gov/pubs/fips/203/final)
- FIPS 204, [Module-Lattice-Based Digital Signature Standard](https://csrc.nist.gov/pubs/fips/204/final)
- FIPS 205, [Stateless Hash-Based Digital Signature Standard](https://csrc.nist.gov/pubs/fips/205/final)

**Comment by hammad93** [comment]
We have a **AOM-TPM-9670V Module** TPM Module

Here is a python module for it, https://tpm2-pytss.readthedocs.io/en/latest/fapi.html


This is a Python library that has already implemented some of the above post quantum cryptography (PQC)
https://github.com/kpdemetriou/pqcrypto/tree/master

If we take a look at the Kyber algorithm at 512 bit strength,
https://github.com/kpdemetriou/pqcrypto/blob/master/sources/kyber512/ref/kem.c

There are multiple ways we can use the TPM.
- Store the PQC private key inside the TPM non-volatile memory
- Generate the key using the TPM's true random number generator (as opposed to pseudo random number generators)

TPM's don't currently support PQC algorithms. Including PQC private key generation into existing TPM modules like the one mentioned is not currently supported. Reference this page for details on supported algorithms, https://trustedcomputinggroup.org/resource/tcg-algorithm-registry/ and [as we can see here](https://trustedcomputinggroup.org/wp-content/uploads/TCG-Algorithm-Registry-Revision-1.34_pub-1.pdf), there are currently no PQC methods. It's likely that the next generation of TPM modules will fully support it. 

In conclusion, it's still very useful to generate the private key on our own and store it on the TPM. The downside of this would be the generation, encryption, and decryption will have to be built but the upside is that the private key is tightly secured. That would remediate security concerns of this repository brought on by https://github.com/hammad93/time_crypt/issues/12

**Comment by hammad93** [comment]
First, we must use the TPM for supported purposes and then explore additional features like PQC.

**Comment by hammad93** [comment]
TPM supported functionality is well understood. PQC algorithms are not available as of TPM 2.0 so this issue can be closed when we analyze whether to use PQC from existing Python libraries.
